sources/metrics/Stride.Metrics.ServerApp/Extensions/MetricDbContextExtensions.cs

```diff
@@ -0,0 +1,128 @@
+        lock (AppIds)
+        {
+            if (!AppIds.TryGetValue(appGuid, out id))
+            {
+                id = -1;
+                var app = context.MetricApps.FirstOrDefault(t => t.AppGuid == appGuid);
+                if (app != null)
+                {
+                    id = app.AppId;
+                    AppIds.Add(appGuid, id);
+                }
+            }
+        }
+        return id;
+    }
+
+    public static int GetOrCreateInstallId(this MetricDbContext context, Guid installGuid)
+    {
+        int id;
+        lock (InstallIds)
+        {
+            if (!InstallIds.TryGetValue(installGuid, out id))
+            {
+                // Create automatically install id on first encounter
+                var install = context.MetricInstalls.FirstOrDefault(t => t.InstallGuid == installGuid);
+
+                if (install == null)
+                {
+                    install = new MetricInstall(installGuid);
+                    context.MetricInstalls.Add(install);
+                    context.SaveChanges();
+                }
+
+                id = install.InstallId;
+                InstallIds.Add(installGuid, id);
+            }
+        }
+        return id;
+    }
+
+    public static int GetMetricId(this MetricDbContext context, Guid metricGuid)
+    {
+        int id;
+        lock (MetricIds)
+        {
+            if (!MetricIds.TryGetValue(metricGuid, out id))
+            {
+                id = -1;
+                var metricDef = context.MetricEventDefinitions.FirstOrDefault(t => t.MetricGuid == metricGuid);
+                if (metricDef != null)
+                {
+                    id = metricDef.MetricId;
+                    MetricIds.Add(metricGuid, id);
+                }
+            }
+        }
+        return id;
+    }
+    internal static MetricEvent SaveNewMetric(this MetricDbContext dbContext, NewMetricMessage newMetric, string ipAddress, DateTime? overrideTime = null, bool disableSave = false)
+    {
+        var applicationId = dbContext.GetApplicationId(newMetric.ApplicationId);
+        if (applicationId == -1)
+        {
+            Trace.TraceError($"Invalid ApplicationId {newMetric.ApplicationId} from {ipAddress}");
+            throw new InvalidOperationException($"Invalid ApplicationId {newMetric.ApplicationId} from {ipAddress}");
+        }
+
+        var installId = dbContext.GetOrCreateInstallId(newMetric.InstallId);
+        var metricId = dbContext.GetMetricId(newMetric.MetricId);
+
+        var metricEvent = new MetricEvent
+        {
+            AppId = applicationId,
+            InstallId = installId,
+            SessionId = newMetric.SessionId,
+            MetricId = metricId,
+            IPAddress = ipAddress,
+            EventId = newMetric.EventId,
+            MetricValue = newMetric.Value ?? string.Empty,
+            Timestamp = overrideTime ?? default
+        };
+
+        var metricAsJson = JsonConvert.SerializeObject(newMetric);
```

Comment:
Would it make sense to move this serialization into the if block - this way we don't do it unnecessarily on the happy path.